docs/guides/connect-to-state-with-url-hash.md

```diff
@@ -43,6 +43,114 @@ export const useBoundStore = create(
 )
 ```
 
-## CodeSandbox Demo
+### CodeSandbox Demo
 
 https://codesandbox.io/s/zustand-state-with-url-hash-demo-f29b88?file=/src/store/index.ts
+
+## Persist and Connect State with URL Parameters (Example: URL Query Parameters)
+
+This example is for when you want to conditionally connect the state to the URL (here we use URL query parameters) when visiting the website from a populated URL, and always want it synced and consistent with something like localstorage.
+
+If you want the URL params to always populate, the conditional check on `window.location.search.slice(1)` can be removed.
+
+The implementation below will update the URL in place, without refresh, as the relevant states change.
+
+```ts
+import { create } from 'zustand'
+import { persist, StateStorage, createJSONStorage } from 'zustand/middleware'
+
+const persistentStorage: StateStorage = {
+  getItem: (key): string => {
+    // Check URL first
+    if (window.location.search.slice(1)) {
+      const searchParams = new URLSearchParams(window.location.search.slice(1))
```

Comment:
```js
      const searchParams = new URLSearchParams(window.location.search.slice(1))
      // consider extracting to a function ----^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

      // so that it could shortened into:
      new URLSearchParams(windowSearchParams())
```